sympy/integrals/integrals.py

```diff
@@ -1242,6 +1242,113 @@ def _sage_(self):
+        >>> from sympy.integrals.integrals import Integral
+        >>> from sympy.calculus.singularities import singularities
+        >>> x = symbols('x')
+
+        >>> g = x + 1
+        >>> Integral(g, (x, -oo, oo)).principal_value()
+        oo
+
+        >>> f = 1/(x**3)
+        >>> singularities(f,x)
+        {0}
+
+        >>> Integral(f, (x, -oo, oo)).principal_value()
+        nan
+
+        This happens so because, the method principal_value returns -oo for Integral from -oo to 0
+        and +oo for Integral from 0 to oo. While adding both the Integrals i.e. -oo + oo we do not
+        get 0 (as we expect) but nan since sympy realises that -oo and +oo aren't actual numbers that
+        can be added/subtracted.
+
+        >>> Integral(f, (x, -oo, 0)).principal_value()
+        -oo
+
+        >>> Integral(f, (x, 0, oo)).principal_value()
+        oo
+
+        >>> Integral(f, (x, 1, 2)).principal_value()
+        3/8
+
+        >>> Integral(f, (x, -2, -1)).principal_value() + Integral(f, (x, 1, 2)).principal_value()
+        0
+
+        >>> Integral(f, (x, 0.4, 3)).principal_value()
+        3.06944444444444
+
+        >>> Integral(f, (x, -3, -0.4)).principal_value() + Integral(f, (x, 0.4, 3)).principal_value()
+        0
+
+
+        References
+        ==========
+
+        .. [1] http://en.wikipedia.org/wiki/Cauchy_principal_value
+        .. [2] http://mathworld.wolfram.com/CauchyPrincipalValue.html
+        """
+        from sympy.calculus import singularities
+
+        x, a, b = self.limits[0]
+
+        r = Dummy('r')
+
+        singularities_list = [s for s in singularities(self.function, x) if s.is_comparable and a < s < b]
+
+        if len(singularities_list) == 0:
+            if a == -oo and b == oo:
+                I = integrate(self.function, (x, -r, r))
+                principle_value = limit(I, r, oo)
+                return principle_value
+            else:
+                I = integrate(self.function, (x, a, b))
+                return I
+
+        if len(singularities_list) == 1:
+            I = integrate(self.function, (x, a, singularities_list[0] - r))
+            principle_value = 0
+            principle_value = principle_value + limit(I, r, 0)
+            I = integrate(self.function, (x, singularities_list[0] + r, b))
+            principle_value = principle_value + limit(I, r, 0)
+            return principle_value
+
+        else:
+
+            for singular_element in range(len(singularities_list)):
+
+                if a < singularities_list[singular_element] < b:
+
+                    if singular_element == 0:
+
+                        I = integrate(self.function, (x, a, singularities_list[singular_element] - r))
+                        principle_value = 0
+                        principle_value = principle_value + limit(I, r, 0)
+                        return principle_value
+
+                    else:
+
+                        I = integrate(self.function, (
```

Comment:
`integrate` is an expensive operation. It should be called only once. That can be achieved if it is used to find an antiderivative. That should be done in the beginning. Then it can also be seen if the integral can be evaluated at all. If not, then the original integral can be returned unevaluated without having to spend time on finding singularities.

Reply:
@jksuom  the `if-else` statements make sure that integral is called only the necessary number of times. Depending upon the number of singularities, the integral will have to be calculated `number of singularities + 1` times (Since that is the definition of cauchy's principal value, we cannot change that) So I think that this cannot be simplified any further.
Forgive me if I misunderstood your remark, what exactly are you suggesting I do?

Reply:
It is not necessary to call `integrate` for each integral. Each one can be computed from a single antiderivative.